File: packages/python/src/tesote_sdk/v2/accounts.py
"""v2 accounts resource: list, get, sync."""

from __future__ import annotations

from typing import Any, Dict, Optional

from ..models import Account, AccountList, AccountSyncStarted
from ..transport import Transport


class AccountsResource:
    """`/v2/accounts` -- list, get, sync."""

    _PREFIX = "/v2/accounts"

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def list(
        self,
        *,
        page: Optional[int] = None,
        per_page: Optional[int] = None,
        include: Optional[str] = None,
        sort: Optional[str] = None,
        cache_ttl: Optional[float] = None,
    ) -> AccountList:
        query: Dict[str, Any] = {
            "page": page,
            "per_page": per_page,
            "include": include,
            "sort": sort,
        }
        response = self._transport.request("GET", self._PREFIX, query=query, cache_ttl=cache_ttl)
        body: Dict[str, Any] = response.json if isinstance(response.json, dict) else {}
        return AccountList.from_dict(body)

    def get(self, account_id: str, *, cache_ttl: Optional[float] = None) -> Account:
        response = self._transport.request(
            "GET", f"{self._PREFIX}/{account_id}", cache_ttl=cache_ttl
        )
        body: Dict[str, Any] = response.json if isinstance(response.json, dict) else {}
        return Account.from_dict(body)

    def sync(
        self,
        account_id: str,
        *,
        idempotency_key: Optional[str] = None,
    ) -> AccountSyncStarted:
        """POST /v2/accounts/{id}/sync. 202 Accepted; idempotency-key header optional."""
        response = self._transport.request(
            "POST",
            f"{self._PREFIX}/{account_id}/sync",
            body={},
            idempotency_key=idempotency_key,
        )
        body: Dict[str, Any] = response.json if isinstance(response.json, dict) else {}
        return AccountSyncStarted.from_dict(body)


__all__ = ["AccountsResource"]

File: packages/ruby/lib/tesote_sdk/v2/accounts.rb
module TesoteSdk
  module V2
    class Accounts
      INDEX_CACHE_TTL = 60
      SHOW_CACHE_TTL = 300

      def initialize(transport)
        @transport = transport
      end

      # GET /v2/accounts
      def list(query = {}, opts: {})
        body = @transport.request('GET', 'accounts', query: query, opts: with_cache(opts, INDEX_CACHE_TTL))
        Models::AccountList.from_hash(body)
      end

      # GET /v2/accounts/{id}
      def get(id, opts: {})
        raise ArgumentError, 'id is required' if id.nil? || id.to_s.empty?

        body = @transport.request('GET', "accounts/#{id}", opts: with_cache(opts, SHOW_CACHE_TTL))
        Models::Account.from_hash(body)
      end

      # POST /v2/accounts/{id}/sync
      # Triggers an async sync; returns SyncStartResult (status: pending).
      def sync(id, opts: {})
        raise ArgumentError, 'id is required' if id.nil? || id.to_s.empty?

        body = @transport.request('POST', "accounts/#{id}/sync", body: {}, opts: opts)
        Models::SyncStartResult.from_hash(body)
      end

      private

      def with_cache(opts, ttl)
        return opts if opts.key?(:cache)

        opts.merge(cache: { ttl: ttl })
      end
    end
  end
end
